drivers/wdf/wdfldr/common.c

```diff
@@ -0,0 +1,826 @@
+
+	if (KeyHandle)
+		ZwClose(KeyHandle);
+
+	return result;
+}
+
+
+NTSTATUS
+NTAPI
+FxLdrQueryUlong(
+	IN HANDLE KeyHandle,
+	IN PUNICODE_STRING ValueName,
+	OUT PULONG Value
+)
+{
+	NTSTATUS status;
+	ULONG resultLength;
+	KEY_VALUE_PARTIAL_INFORMATION keyValue;
+
+	keyValue.DataLength = 0;
+	keyValue.TitleIndex = 0;
+	keyValue.Type = 0;
+	keyValue.Data[0] = 0;
+	status = ZwQueryValueKey(KeyHandle, ValueName, KeyValuePartialInformation, &keyValue, sizeof(keyValue), &resultLength);
+
+	if (NT_SUCCESS(status)) 
+	{
+		if (keyValue.Type != REG_DWORD || keyValue.DataLength != 4) 
+		{
+			status = STATUS_INVALID_BUFFER_SIZE;
+		}
+		else 
+		{
+			*Value = keyValue.Data[0];
+			status = STATUS_SUCCESS;
+		}
+	}
+	else 
+	{
+		if (WdfLdrDiags) 
+		{
+			DbgPrint("WdfLdr: FxLdrQueryUlong - ");
+			DbgPrint("ERROR: ZwQueryValueKey failed with Status 0x%x\n", status);
+		}
+	}
+
+	return status;
+}
+
+
+NTSTATUS
+NTAPI
+FxLdrQueryData(
+	IN HANDLE KeyHandle,
+	IN PUNICODE_STRING ValueName,
+	IN ULONG Tag,
+	OUT PKEY_VALUE_PARTIAL_INFORMATION* KeyValPartialInfo
+)
+{
+	PKEY_VALUE_PARTIAL_INFORMATION pKeyInfo;
+	NTSTATUS status;
+	ULONG resultLength;
+
+	*KeyValPartialInfo = NULL;
+	for (;;)
+	{
+		status = ZwQueryValueKey(KeyHandle, ValueName, KeyValuePartialInformation, NULL, 0, &resultLength);
+		if (status != STATUS_BUFFER_TOO_SMALL) 
+		{
+			if (!NT_SUCCESS(status) && WdfLdrDiags) 
+			{
+				DbgPrint("WdfLdr: FxLdrQueryData - ");
+				DbgPrint("ERROR: ZwQueryValueKey failed with status 0x%x\n", status);
+			}
+
+			return status;
+		}
+
+		status = RtlULongAdd(resultLength, 0xCu, &resultLength);
+		if (!NT_SUCCESS(status)) 
+		{
+			if (WdfLdrDiags) 
+			{
+				DbgPrint("WdfLdr: FxLdrQueryData - ");
+				DbgPrint("ERROR: Computing length of data under %wZ failed with status 0x%x\n", ValueName, status);
+			}
+
+			return status;
+		}
+
+		pKeyInfo = ExAllocatePoolWithTag(PagedPool, resultLength, Tag);
+
+		if (pKeyInfo == NULL)
+		{
+			break;
+		}
+
+		memset(pKeyInfo, 0, resultLength);
+		status = ZwQueryValueKey(
+			KeyHandle,
+			ValueName,
+			KeyValuePartialInformation,
+			pKeyInfo,
+			resultLength,
+			&resultLength);
+
+		if (NT_SUCCESS(status)) 
+		{
+			*KeyValPartialInfo = pKeyInfo;
+			return status;
+		}
+
+		ExFreePoolWithTag(pKeyInfo, 0);
+
+		if (status != STATUS_BUFFER_TOO_SMALL) 
+		{
+			if (WdfLdrDiags) {
+				DbgPrint("WdfLdr: FxLdrQueryData - ");
+				DbgPrint("ERROR: ZwQueryValueKey (%wZ) failed with Status 0x%x\n", ValueName, status);
+			}
+
+			return status;
+		}
+	}
+
+	if (WdfLdrDiags) 
+	{
+		DbgPrint("WdfLdr: FxLdrQueryData - ");
+		DbgPrint("ERROR: ExAllocatePoolWithTag failed with Status 0x%x\n", STATUS_INSUFFICIENT_RESOURCES);
+	}
+
+	return STATUS_INSUFFICIENT_RESOURCES;
+}
+
+PWCHAR
+FreeString(
+	IN PUNICODE_STRING String
+)
+{
+	PWCHAR buffer;
+	buffer = String->Buffer;
+
+	if (buffer) 
+	{
+		ExFreePoolWithTag(buffer, 0);
+		buffer = 0;
+		String->Length = 0;
+		String->Buffer = 0;
+	}
+
+	return buffer;
```

Comment:
What's the point? Just return NULL, that's what you'll return anyway (be it freed or not).
Actually, what's the point of returning something?